tests/virt_autotest/vgpu.pm

```diff
@@ -0,0 +1,368 @@
+
+    # Choose a ramdom vgpu type and create a vGPU
+    my $vgpu_type_name = ${avail_types [int(rand($#avail_types + 1))]};
+    my $vgpu_type_id = script_output("grep -l '${vgpu_type_name}' */name | cut -d '/' -f1");
+    my $vgpu_id = script_output('uuidgen');
+    assert_script_run("echo ${vgpu_id} > ${vgpu_type_id}/create");
+
+    # Verify if vGPU created successfully
+    record_info("vGPU created", "$vgpu_type_name " . script_output("lsmdev"));
+    assert_script_run("dmesg | tail");
+    assert_script_run("nvidia-smi vgpu -q");
+
+    return ${vgpu_id};
+}
+
+sub assign_vgpu_to_guest {
+    my ($uuid, $vm, $slot) = @_;
+
+    # The UEFI guest created in virt test is ok for vGPU test
+    die "vGPU only works on UEFI guests!" unless $vm =~ /uefi/i;
+    assert_script_run("virsh shutdown $vm") unless script_output("virsh domstate $vm") eq "shut off";
+    # Save guest xml to /tmp/vm.xml, undefine the current one and define with the new xml
+    my $vm_xml_file = "/tmp/$vm.xml";
+    assert_script_run "virsh dumpxml --inactive $vm > $vm_xml_file";
+    assert_script_run "virsh undefine $vm";
+
+    # Add the vgpu device section to guest configuration file
+    die "PCI slot '0x0a' has already been used" if script_run("grep \"slot='$slot'\" $vm_xml_file") == 0;
+    my $vgpu_xml_section = "<hostdev mode='subsystem' type='mdev' model='vfio-pci' display='off'>\\\n  <source>\\\n    <address uuid='$uuid'/>\\\n  </source>\\\n  <address type='pci' domain='0x0000' bus='0x00' slot='$slot' function='0x0'/>\\\n</hostdev>";
+
+    assert_script_run("sed -i \"/<devices>/a\\\\${vgpu_xml_section}\" $vm_xml_file");
+    upload_logs($vm_xml_file);
+    assert_script_run "virsh define $vm_xml_file";
+    assert_script_run "virsh start $vm";
+    wait_guest_online($vm);
+    assert_script_run "ssh root\@$vm 'lspci | grep NVIDIA'";
+    record_info("vGPU attached to $vm", script_output("ssh root\@$vm 'lspci | grep NVIDIA'"));
+}
+
+sub install_vgpu_grid_driver {
+    my ($vm, $driver_url) = @_;
+
+    # Download drivers from fileserver
+    die "The vGPU driver URL requires to be in 'http://...' format!" unless $driver_url =~ /http:\/\/.*\/(.*\.run)/;
+    my ${driver_file} = $1;
+
+    # Check if the driver has been already installed
+    if (script_run("ssh root\@$vm 'nvidia-smi'") == 0) {
+        my $grid_version = script_output("ssh root\@$vm 'nvidia-smi -q | grep \"Driver Version\" | grep -oE \"[0-9.]+\"'");
+        if (${driver_file} =~ /$grid_version/) {
+            record_info("Warning", "vGPU grid driver $grid_version has already been loaded!", result => 'softfail');
+            return;
+        }
+    }
+
+    # Install dependencies seperately. It is easier to locate problem this way.
+    download_script(${driver_file}, script_url => ${driver_url}, machine => $vm);
+    assert_script_run "ssh root\@$vm 'zypper -n in kernel-default-devel'";
+    assert_script_run "ssh root\@$vm 'zypper -n in libglvnd-devel'";
```

Comment:
Why do not let them in one line:
`assert_script_run "ssh root\@$vm 'zypper -n in kernel-default-devel libglvnd-devel '";`

Reply:
I did it on purpose to narrow down problems which might happen. When there are multiple packages in "zypper in", their dependencies are mixed up. It's hard to tell which package is problematic, in particular for reporting a bug. I am not very certain with my words but I do find these words in my test notes, `Do install the packages seperately in automation test even if they are installed together in SUSE DOC`